File: 74-search-a-2d-matrix/solution.cpp
#include <vector>

class Solution
{
public:
  bool searchMatrix(std::vector<std::vector<int>> &matrix, int target)
  {
    int left = 0, right = matrix.size() - 1;
    while (left <= right)
    {
      const int mid = (left + right) / 2;
      const int num = matrix[mid][0];
      if (num == target)
      {
        return true;
      }

      if (target < num)
      {
        right = mid - 1;
      }
      else
      {
        left = mid + 1;
      }
    }

    if (left >= matrix.size() || matrix[left][0] > target)
    {
      --left;
    }

    if (left < 0)
    {
      return false;
    }

    const auto &row = matrix[left];
    left = 0, right = row.size() - 1;
    while (left <= right)
    {
      const int mid = (left + right) / 2;
      const int num = row[mid];
      if (num == target)
      {
        return true;
      }

      if (target < num)
      {
        right = mid - 1;
      }
      else
      {
        left = mid + 1;
      }
    }

    return false;
  }
};

File: 74-search-a-2d-matrix/solution.py
class Solution:
  def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
    left, right = 0, len(matrix) - 1
    while left <= right:
      mid = (left + right) // 2
      num = matrix[mid][0]
      if num == target:
        return True

      if target < num:
        right = mid - 1
      else:
        left = mid + 1

    if left >= len(matrix) or matrix[left][0] > target:
      left -= 1

    if left < 0:
      return False

    row = matrix[left]
    left, right = 0, len(row) - 1
    while left <= right:
      mid = (left + right) // 2
      num = row[mid]
      if num == target:
        return True

      if target < num:
        right = mid - 1
      else:
        left = mid + 1

    return False
